recipe/2024-04-terraform-provider-secrets.md

```diff
@@ -0,0 +1,260 @@
+
+### Detailed Design
+
+The FindSecretIds function in the DriverWithSecrets interface is updated to return a map of SecretStoreIds and keys. These will hold data for a Terraform module as before and in addition, also hold SecretStoreIds and keys for provider configuration and environment variables, iterating through the input environment recipe configuration. 
+
+The LoadSecrets method in the SecretLoader interface will be updated return secret data for multiple Ids returned from FindSecretIds().
+
+#### Advantages (of each option considered)
+Retrieving secrets in the Engine and passing them into the Driver was discussed positively as a way forward. Looking ahead with Containerization, the Driver will have all the information it needs to be able to be spin off into a separate process.
+
+#### Disadvantages (of each option considered)
+N/A
+
+#### Proposed Option
+N/A
+
+### API design (if applicable)
+N/A
+
+### CLI Design (if applicable)
+N/A
+
+### Implementation Details
+###Recipes (Engine and Driver)
+
+
+Driver Updates:
+```go
+//*** CURRENT ***//
+
+type DriverWithSecrets interface {
+  ...
+	// FindSecretIDs gets the secret store resource ID references associated with git private terraform repository source.
+	// In the future it will be extended to get secret references for provider secrets.
+	FindSecretIDs(ctx context.Context, config recipes.Configuration, definition recipes.EnvironmentDefinition) (string, error)
+}
+
+//*** NEW ***//
+
+// FindSecretIDs will be updated in current DriverWithSecrets interface to now return map of secretStoreIds and keys
+type DriverWithSecrets interface {
+  ...
+	// FindSecretIDs gets the secretStore resource IDs and keys for the recipe including module, provider, envSecrets.
+	FindSecretIDs(ctx context.Context, config recipes.Configuration, definition recipes.EnvironmentDefinition) (map[string]string, error)
+}
+```
+
+Engine Updates
+
+```go
+//*** CURRENT ***//
+
+type SecretsLoader interface {
+	LoadSecrets(ctx context.Context, secretStore string) (v20231001preview.SecretStoresClientListSecretsResponse, error)
+}
+
+// BaseOptions is the base options for the driver operations.
+type BaseOptions struct {
+  ...
+
+	// Secrets specifies the module authentication information stored in the secret store.
+	Secrets v20231001preview.SecretStoresClientListSecretsResponse
+}
+
+
+//*** NEW ***//
+
+// LoadSecrets function in SecretsLoader interface will populate resolved secrets based on the Ids returned from FindSecretIDs(). The result is passed as BaseOptions.Secrets to the Driver in Execute() call.
+type SecretsLoader interface {
+	LoadSecrets(map[string]string, ...)(map[string]map[string]string, error){ // returns secretStoreId[key]value
+}
+
+// When calling driver.Execute we pass in BaseOptions which is structured as:
+type BaseOptions struct {
+   ...
+
+	// Secrets specifies the module authentication information stored in the secret store.
+	Secrets map[string]map[string]string
```

Comment:
It would be great to call out the keys and values in the docs (and in the review).

eg:

> Secrets is a map from ??? to ???

Reply:
updated document